### PR TITLE
Added some CICD best practices to the documentation

### LINKED ISSUE #4685: [Docs] document ci/cd best practices
### Description

document some general ci/cd best practices discussed by Bernhard in this conversation:: https://discuss.flyte.org/t/15853851/hi-team-i-d-like-to-understand-a-little-more-about-best-prac#1eaa0ca8-d3ec-49a9-9f46-b8537c610cb3

### Are you sure this issue hasn't been raised already?

- [X] Yes

### Have you read the Code of Conduct?

- [X] Yes

### PR DESCRIPTION
**Pull request by Murdock9803**
## Tracking issue
Closes #4685

## Why are the changes needed?
The documentation at [this page](https://docs.flyte.org/en/latest/user_guide/flyte_fundamentals/registering_workflows.html#ci-cd-with-flyte-and-github-actions) talks about Github Actions with CI/CD and Flyte. But, It can be new to many people to work with CICD process and some extra documentation will be helpful.

## What changes were proposed in this pull request?
In the [docs/user_guide/flyte_fundamentals/registering_workflows.md](https://github.com/flyteorg/flyte/blob/master/docs/user_guide/flyte_fundamentals/registering_workflows.md) file, which can be seen deployed at [this page](https://docs.flyte.org/en/latest/user_guide/flyte_fundamentals/registering_workflows.html#ci-cd-with-flyte-and-github-actions), I added some documentation related to CI/CD best practices near the end of the file.

## How was this patch tested?
The Pull Request is about simple Documentation changes.

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

**Comment by Murdock9803** [comment]
@davidmirror-ops @samhita-alla I tried to add the documentation regarding CICD, please have a look at this PR 🙌

**Comment by Murdock9803** [comment]
I merged the changes (done by recently merged PRs) into my branch, so this PR does not have any merge conflicts 🙌

**Comment by Murdock9803** [comment]
@davidmirror-ops The checks have passed, Please have a look at the Documentation and suggest required changes 🙌

File: docs/user_guide/flyte_fundamentals/registering_workflows.md [hunk 1]
```diff
@@ -358,6 +358,34 @@ two GitHub actions that facilitates this:
   of Flyte packages, for example, the `.tgz` archives that are created by
   `pyflyte package`.
 
+### Some CI/CD best practices
+
+Several best practices are essential for ensuring smooth versioning, secure automation and image management. Tools like Docker, Helm and Github Actions help teams streamline their build, tagging, and release processes while maintaining security and scalability.
```
Comment (samhita-alla) [inline docs/user_guide/flyte_fundamentals/registering_workflows.md:363]: @davidmirror-ops should the tone sound like we're just repeating what Bernhard suggested? i don't think these should come across as if we're suggesting, right?

Comment (davidmirror-ops) [reply docs/user_guide/flyte_fundamentals/registering_workflows.md:363]: @samhita-alla Agree. 
@RaghavMangla the essence of the issue is to generally capture what was suggested in the conversation. I recognize someone needs significant context to do so. The antigoals are to repeat what Bernhard said, making it sound as a requirement. Almost every organization will have a different way of doing CI/CD. 
Also try to not overemphasize Docker, as Flyte supports any other CRI-compliant container engine.

File: docs/user_guide/flyte_fundamentals/registering_workflows.md [hunk 2]
```diff
@@ -358,6 +358,34 @@ two GitHub actions that facilitates this:
...
+
+#### Best practices to follow
+
+- **GitHub Actions for Workflow Automation**: Leverage GitHub Actions to automate the release process, including critical parts such as generating git tags, Docker images, and releasing Helm and manifest files. Automating these workflows reduces manual errors and speeds up the deployment cycle.
+
+- **Docker for Image Management**:
+Use Docker to build, tag, and push images based on the version of the build. These Docker images are integral to the deployment pipeline, allowing Flyte workflows and other tasks to reference the correct image without manual work.
+
+- **Parallel Execution with Matrix Strategy**:
+To optimize performance and efficiency through parallel execution of tasks, employ a matrix strategy in GitHub Actions. This is used in the `publish-flyte-component-image` job.
```
Comment (davidmirror-ops) [inline docs/user_guide/flyte_fundamentals/registering_workflows.md:381]: This is not exactly clear.
Also, I'm not opposed to assisting technical writing with AI but we need to be careful of not relying too much on generated content for docs writing.

Comment (Murdock9803) [reply docs/user_guide/flyte_fundamentals/registering_workflows.md:388]: I agree, @davidmirror-ops. I just needed some more context which I got from the review comment above. I will definitely try to come up with better text that's aimed.

**Comment by Murdock9803** [comment]
@samhita-alla @davidmirror-ops I tried to improve the documentation, Please have a look. As this is a documentation change, I believe continuous feedback and improvement can make for a good article 😄 🙌

File: docs/user_guide/flyte_fundamentals/registering_workflows.md [hunk 3]
```diff
@@ -365,6 +365,21 @@ two GitHub actions that facilitates this:
...
+
+- **Workflow Serialization and Registration** : Workflows should be serialized and registered based on the versioning of the build and the container image. Depending on whether the build is for a feature branch or main, the registration domain should be adjusted accordingly. For more context, please visit the [Registering workflows](https://docs.flyte.org/en/latest/user_guide/flyte_fundamentals/registering_workflows.html) page.
+
+- **Container Image Specification** : When managing multiple images across tasks within a Flyte workflow, use the `--image` flag during registration to specify which image to use. This avoids hardcoding the image within the task definition, promoting reusability and flexibility in workflows.
+
+- **Helm and GoReleaser for Packaging and Releasing**: Use Helm for managing Kubernetes package releases and GoReleaser for automating the release of Go applications. These tools streamline the packaging and distribution processes, making the CI/CD pipeline more robust and manageable. These are used in the `helm-release` and `manifest-release` jobs.
```
Comment (davidmirror-ops) [inline docs/user_guide/flyte_fundamentals/registering_workflows.md:380]: I think we can remove this one, we don't expect users to release Go apps with Flyte.

File: docs/user_guide/flyte_fundamentals/registering_workflows.md [hunk 4]
```diff
@@ -365,6 +365,21 @@ two GitHub actions that facilitates this:
   of Flyte packages, for example, the `.tgz` archives that are created by
   `pyflyte package`.
 
+### Some CI/CD best practices
+
+In case Flyte workflows are registered on each commit in your build pipelines, you can consider the following recommendations and approach:
+
+- **Versioning Strategy** : Determining the version of the build for different types of commits makes them consistent and identifiable. For commits on feature branches, use `<branch-name>-<short-commit-hash>` and for the ones on main branches, use `main-<short-commit-hash>`. Use version numbers for the released (tagged) versions.
+
+- **Container Image Management** : With Docker or any other CRI-compliant engine, build and push images to maintain traceability and consistency across environments.
```
Comment (davidmirror-ops) [inline docs/user_guide/flyte_fundamentals/registering_workflows.md:374]: Also we can remove this one as it's pretty generic

**Review by davidmirror-ops — CHANGES_REQUESTED** [review]
Left some comments. Thanks!

File: docs/user_guide/flyte_fundamentals/registering_workflows.md (hunk 3 above)
Comment (Murdock9803) [reply docs/user_guide/flyte_fundamentals/registering_workflows.md:380]: Sure 🙌

File: docs/user_guide/flyte_fundamentals/registering_workflows.md (hunk 4 above)
Comment (Murdock9803) [reply docs/user_guide/flyte_fundamentals/registering_workflows.md]: I agree 🙌

**Comment by Murdock9803** [comment]
@davidmirror-ops I removed the unnecessary lines, please suggest further changes If any 🙌 
Should I add more info about some particular topics ?

**Review by davidmirror-ops — APPROVED** [review]
Thank you

**Comment by Murdock9803** [comment]
@davidmirror-ops Thanks for approving the changes ❤️. I have pulled in the latest changes from the main branch into my branch so its up to date 🙌

**Review by davidmirror-ops — APPROVED** [review]

**Pull request merged**